providers/dns/script/script.go

```diff
@@ -0,0 +1,75 @@
+// Package script implements a manual DNS provider which runs a script for
+// adding/removing the DNS record.
+//
+// The file name of the external script is specified in the environment
+// variable SCRIPT_PATH. When it is run by lego, three command-line parameters
+// are passed to it: The action ("add" or "remove"), the fully-qualified domain
+// name, the value for the record and the TTL.
+//
+// For example, requesting a certificate for the domain 'foo.example.com' can
+// be achieved by calling lego as follows:
+//
+//  SCRIPT_PATH=./update-dns.sh \
+//    lego --dns script \
+//    --domains foo.example.com \
+//    --email invalid@example.com run
+//
+// It will then call the script './update-dns.sh' with like this:
+//
+//  ./update-dns.sh "add" "_acme-challenge.foo.example.com." "MsijOYZxqyjGnFGwhjrhfg-Xgbl5r68WPda0J9EgqqI" "120"
+//
+// The script then needs to make sure the record is inserted. When it returns
+// an error via a non-zero exit code, lego aborts.
+//
+// When the record is to be removed again, the script is called with the first
+// command-line parameter set to "remove" instead of "add".
+package script
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/xenolf/lego/acme"
+)
+
+// DNSProvider adds and removes the record for the DNS challenge by calling a
+// script with command-line parameters.
+type DNSProvider struct {
+	script string
+}
+
+// NewDNSProvider returns a new DNS provider which runs the script in the
+// environment variable SCRIPT_PATH for adding and removing the DNS record.
+func NewDNSProvider() (*DNSProvider, error) {
+	s := os.Getenv("SCRIPT_PATH")
+	if s == "" {
+		fmt.Printf("error, script not set")
+		return nil, errors.New("environment variable SCRIPT_PATH not set")
+	}
+
+	fmt.Printf("init dns provider with script %v", s)
+	return &DNSProvider{script: s}, nil
+}
+
+// Present creates a TXT record to fulfil the dns-01 challenge.
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl := acme.DNS01Record(domain, keyAuth)
+	cmd := exec.Command(d.script, "add", fqdn, value, strconv.Itoa(ttl))
```

Comment:
The argument value for the action should probably be "present" instead of "add"

Reply:
Yeah, I thought about that. From the point of view of a user, it's not clear why it's `present`. I had to look it up to. Since the script/program is called to "add" or "modify" a record, I thought it'd be clearer to the script/program author to call it "add".

Happy to change it though ;)

Reply:
So, let me know what you think.